---
title: "R Notebook"
output: html_notebook
---

```{r}
library("tidyr")
library("plyr")
library('ggplot2')
library('dplyr')
library("glue")
library("ggVennDiagram")

wkdir = "~/Desktop/GitHub/Obesity/NewExtractions/H9N2"
setwd(wkdir)
savedir = "~/Desktop/GitHub/Obesity/NewExtractions/H9N2/Output_Figures"

source("~/Desktop/Obesity/OLD/FromKate/FD_functions.R")
```

# Pre Sequence Analysis
Load Data
```{r}
df = read.csv("NewH9N2Seq.csv", header = TRUE, sep = ",")
df$Ferret_Time = paste0(df$Ferret.ID,"_",df$Timepoint)
df$M.gene.Ct = as.numeric(df$M.gene.Ct)
df$Timepoint = factor(df$Timepoint, levels = c("2dpi","4dpi","6dpi","8dpi","10dpi","12dpi"))
```

```{r}
x = ggplot(df, aes(x = Ferret_Time, y = M.gene.Ct, color = Good.Gel)) +
  geom_point() +
  #facet_grid(~Diet)
  #facet_grid(~Role)
  PlotTheme1
print(x)
ggsave("CT_ferret.png", x, height = 5, width = 7.5, path = savedir)
```

```{r}
ggplot(df, aes(x = Timepoint, y = M.gene.Ct)) +
  geom_point()
```

```{r}
diets = read.csv("Diets.csv", header = TRUE, sep = ",")
```

```{r}
#flu_align = read.csv("flu.alignment.csv", header = TRUE, sep = ",")
#flu_align$TotalReads = as.numeric(flu_align$TotalReads)
```

# Sequencing Analysis

Specifying thresholds and plotting variables
```{r}
diet = c("Obese","Lean","Control")
dietColors = c("#FF9933","#66CCFF","#606060")
names(dietColors) = diet
DietcolScale_fill <- scale_fill_manual(name = "grp",values = dietColors)
DietcolScale <- scale_colour_manual(name = "grp",values = dietColors)

cov_cut = 200
freq_cut = 0.01
pvalcut  = 0.05

ntlist = c("A","C","G","T")
SEGMENTS = c('H9N2_PB2','H9N2_PB1','H9N2_PA','H9N2_HA','H9N2_NP','H9N2_NA','H9N2_MP','H9N2_NS')
INFECTIONS = c("Index","Contact","Aerosol","Control")
```

Loading in coverage file & segment size information
```{r}
cov = read.csv("./avg_coverage/H9N2.coverage.csv", header = TRUE, sep = ",")

seg_sizes = "SegmentSize.csv"
sizes = read.csv(file=seg_sizes,header=T,sep=",",na.strings = c(''))
GenomeSize = (sizes %>% filter(segment == 'H9N2_GENOME'))$SegmentSize

cov$segment = factor(cov$segment, levels = SEGMENTS)
```

Checking if data passes thresholds & make coverage plots
```{r}
cov_check = CoverageAcross(cov,cov_cut,40,sizes, wkdir)
```

Loading in variant files & metadata
```{r}
varfile = "./varfiles/H9N2.VariantsOnly.0.01.200.csv"
metafile = "NEW_Metadata.csv"

# read and rearrange the data
vars = read.csv(file=varfile,header=T,sep=",",na.strings = c(''))
vars$name = vars$sample

meta = read.csv(file=metafile,header=T,sep=",",na.strings = c(''))
meta = merge(meta, cov_check, by.x = c("sample"), by.y = c("name"))

meta$Ct_Mgene = as.numeric(meta$Ct_Mgene)
```

Rearranging variant dataframe
```{r}
vdf = ArrangeVarWRep(vars)
# already have replicate data in the varfiles from running CompareReps.v2.py script
vdf = vdf[!duplicated(vdf), ] %>% droplevels()
nrow(vdf)
```

Filtering variant df by timo binocheck
```{r}
vdf$binocheck = factor(vdf$binocheck, levels = c("False","R1","R2","True"))
vdf_bino = filter(vdf, binocheck == "True")
vdf_bino = vdf_bino[!duplicated(vdf_bino), ] %>% droplevels()
nrow(vdf_bino)
# this really gets rid of a lot of variants (~1000)

vdf_nobino = filter(vdf, binocheck != "True")
vdf_nobino = vdf_nobino[!duplicated(vdf_nobino), ] %>% droplevels()
nrow(vdf_nobino)

range(vdf_nobino$minorfreq)
ggplot(vdf_nobino, aes(x = minorfreq)) +
  geom_histogram(binwidth = 0.01) +
  PlotTheme1
```

Filtering variant df with frequency cutoffs
```{r}
nrow(vdf_bino)
vdf = filter(vdf_bino, minorfreq1 >= freq_cut & 
               minorfreq2 >= freq_cut & 
               minor %in% ntlist) %>% 
            droplevels()
# based on MAF study, reps and 0.01% cutoff was best combo
#filter each replicate separately rather than using the average

vdf = vdf[!duplicated(vdf), ] %>% droplevels()
nrow(vdf)
# does not eliminate any variants here
```

Filtering variant df by metadata
```{r}
vdf = merge(vdf,meta, by = c("sample","segment"))
vdf = filter(vdf, quality == "good")
vdf = vdf[!duplicated(vdf), ] %>% droplevels()

vdf$segment = factor(vdf$segment, levels = SEGMENTS)
vdf$inf_route = factor(vdf$inf_route, levels = INFECTIONS)
```

Tallying number of ferrets with variants
```{r}
fercount = select(vdf,ferretID,DPI,diet,inf_route)
fercount = fercount[!duplicated(fercount), ]  %>% 
  unique() %>% 
  group_by(diet,inf_route,DPI) %>% 
  tally()

# Counting the number of ferrets with variants, just by ferretID not DPI
filter(vdf, diet == "Lean" & inf_route == "Index") %>% count(ferretID)
filter(vdf, diet == "Obese" & inf_route == "Index") %>% count(ferretID)
filter(vdf, diet == "Lean" & inf_route == "Contact") %>% count(ferretID)
filter(vdf, diet == "Obese" & inf_route == "Contact") %>% count(ferretID)
```

Plotting ferret tally
```{r}
ggplot(fercount, aes(x = DPI, y = n, fill = diet)) +
  geom_col(position = "stack") +
  facet_grid(~inf_route) +
  ylab("Number of Ferrets with SNVs") +
  PlotTheme3 +
  DietcolScale_fill
```

Consensus changes
```{r}
con_change = filter(vdf, stocknt != major) %>%
  filter(major %in% ntlist)
con_change = con_change[!duplicated(con_change), ]
nrow(con_change)

con_change$maj = paste0(con_change$segment,"_",con_change$stock,con_change$ntpos,con_change$major)
con_change$ferretID_maj = paste0(con_change$ferretID,"_",con_change$maj)
con_change = con_change[!duplicated(con_change$ferretID_maj),] 
# not counting same consensus change but just on different days - basically counting unique consensus changes

nrow(con_change)
cons = count(con_change,ferretID,diet,inf_route)

fers = c("2232","2231","1914","1794","1409","1980","1986","1797","2254","2251")
cons$ferretID = factor(cons$ferretID, levels = fers)
cons$inf_route = factor(cons$inf_route, levels = c("Index","Contact"))

ConChange = ggplot(cons, aes(x = ferretID, y = n, fill = diet)) + 
  geom_col() +
  facet_grid(~inf_route) +
  PlotTheme1 +
  DietcolScale_fill
print(ConChange)
ggsave(ConChange, file = "ConChange.pdf", path = savedir)

select(con_change, stocknt, major, minor)
# in all cases, the stocknt is the minor -> has been replaced by another nt
# did these arise as minors first?
```

Plotting consensus changes
```{r}
ggplot(con_change, aes(x = DPI, y = majorfreq, color = diet)) +
  geom_point() +
  facet_grid(inf_route~ferretID) +
  PlotTheme1 +
  DietcolScale

group_by(con_change, diet, inf_route,ferretID) %>%
  tally()
```

Can the consensus changes be detected as minor variants first?
```{r}
majpos = c("1409_H9N2_NP_326","1409_H9N2_PB1_1882","1794_H9N2_HA_284","1980_H9N2_PB1_965",
           "1986_H9N2_NS_652","1986_H9N2_PA_1483","2232_H9N2_HA_747","2232_H9N2_PB2_639",
           "2232_H9N2_PB1_1882")

minorvdf = filter(vdf, stocknt == major)
minorvdf = minorvdf[!duplicated(minorvdf), ]
nrow(minorvdf)
  
minorvdf$con_position = paste0(minorvdf$ferretID,"_",minorvdf$segment,"_",minorvdf$ntpos)
pos_minors = filter(minorvdf, con_position %in% majpos) 
pos_minors$freq = pos_minors$minorfreq
pos_minors = pos_minors[!duplicated(pos_minors), ] %>% droplevels()

pos_majors = con_change %>% select(!c("maj","ferretID_maj"))
pos_majors$con_position = paste0(pos_majors$ferretID,"_",pos_majors$segment,"_",pos_majors$ntpos)
pos_majors$freq = pos_majors$majorfreq

all_pos = rbind(pos_majors,pos_minors) %>% droplevels()
all_pos$segment = factor(all_pos$segment, levels = SEGMENTS)
all_pos$diet = factor(all_pos$diet, levels = c("Obese","Lean"))

ggplot(all_pos, aes(x = DPI, y = freq, color = con_position)) +
  geom_point() +
  #geom_point(aes(y = minorfreq)) +
  geom_line(aes(group = con_position)) +
  facet_grid(~ferretID) +
  PlotTheme1
```

Tallying SNVs
```{r}
# can make these groupings whatever you want

# count the number of SNVs per sample
group_list_seg = c('ferretID','segment',"DPI","diet","inf_route","cohort") # counts across each segment 
group_list_gen = c('ferretID',"DPI","diet","inf_route","cohort") # Counts across entire genome

seg_count = TallyIt(minorvdf, group_list_seg, "snv_count")
gen_count = TallyIt(minorvdf, group_list_gen, "snv_count") 

# Average Number of Variants per Sample
gen_count_avg = group_by(gen_count, DPI, diet, inf_route) %>%
  mutate(avgSNV = mean(snv_count), sdSNV = sd(snv_count))
```

Calculating Shannon Entropy
```{r}
minorvdf = ShannonPos(minorvdf)
minorvdf$SegmentSize = as.numeric(minorvdf$SegmentSize)
minorvdf$shannon_perkb = (minorvdf$segment_shan/(minorvdf$SegmentSize/1000))
minorvdf$normalized_shannon = (minorvdf$shannon/GenomeSize)

# shannon_ntpos = shannon entropy at that nt pos - should always be between 0 and 1 for each sample
# segment_shan = sum of all nt_pos per segment for each sample
# shannon = sum of all segment_shan across genome for each sample
# shannon_perkb = segment shannon per kb (segment specific) for ach sample
# normalized_shannon = shannon divided by genome size (can make per kb by dividing by 1000) for each sample

# Average Shannon Entropy per Site per Sample (using normalized_shannon)
shan_g = select(minorvdf, ferretID, DPI, diet, inf_route, normalized_shannon) 
shan_g = shan_g[!duplicated(shan_g),]
shan_g_avg = group_by(shan_g, DPI, diet, inf_route) %>% 
  mutate(avgShan = mean(normalized_shannon), sdShan = sd(normalized_shannon))

shan_segkb = select(minorvdf, ferretID, DPI, diet, inf_route, segment, shannon_perkb)
shan_segkb = shan_segkb[!duplicated(shan_segkb),]
shan_segkb_avg = group_by(shan_segkb, DPI, diet, inf_route, segment) %>% 
  mutate(avgShan = mean(shannon_perkb), sdShan = sd(shannon_perkb))
```

Making plots for variant tallies and Shannon entropy 
```{r}
VarPlot = ggplot(gen_count_avg, aes(x=DPI, fill = diet)) +
  geom_col(aes(y = avgSNV, group = diet), position = "dodge") +
  geom_jitter(aes(group = diet, y = snv_count), width  = 0.15) +
  geom_errorbar(aes(ymin = avgSNV - sdSNV,
                    ymax = avgSNV + sdSNV)) +
  ylab("Average number of SNVs per sample") +
  facet_grid(~inf_route+diet) +
  PlotTheme1 +
  DietcolScale_fill
print(VarPlot)
ggsave(VarPlot, file = "VariantCount.pdf", path = savedir)

ShannonPlot = ggplot(shan_g_avg, aes(x=DPI, fill = diet)) +
  geom_col(aes(y = avgShan, group = diet), position = "dodge") + #plots average
  geom_jitter(aes(y = normalized_shannon, group = diet), width  = 0.15) + #plots value for each sample
  geom_errorbar(aes(ymin = avgShan - sdShan,
                    ymax = avgShan + sdShan)) +
  ylab("Average Shannon entropy per site across genome") +
  facet_grid(~inf_route+diet) +
  PlotTheme1 +
  DietcolScale_fill
print(ShannonPlot)
ggsave(ShannonPlot, file = "MeanShanPerSite.pdf", path = savedir)

ShannonPlot2 = ggplot(shan_segkb_avg, aes(x=DPI, fill = diet)) +
  geom_col(aes(y = avgShan, group = diet), position = "dodge") + # plots average
  #geom_point(aes(y = shannon_perkb, group = diet)) + # plots value for each segment, sample
  #geom_errorbar(aes(ymin = avgShan - sdShan,
  #                  ymax = avgShan + sdShan)) +
  ylab("Average Shannon entropy per kB across segments") +
  facet_grid(~segment) +
  PlotTheme1 +
  DietcolScale_fill
print(ShannonPlot2)
ggsave(ShannonPlot2, file = "MeanShanPerSegKB.pdf", path = savedir)
```

```{r}
IndexSNVs_perFerret = ggplot(filter(gen_count_avg, inf_route == "Index"),
                             aes(x = DPI, y = snv_count, color = diet, shape = cohort)) +
  geom_point() +
  geom_line(aes(group = ferretID)) +
  geom_label(aes(label = ferretID)) +
  facet_grid(~diet+inf_route)+ 
  PlotTheme1 +
  DietcolScale
print(IndexSNVs_perFerret)
ggsave("IndexSNVs_perFerret.png", IndexSNVs_perFerret, path = savedir)

ContactSNVs_perFerret = ggplot(filter(gen_count_avg, inf_route == "Contact"),
                              aes(x = DPI, y = snv_count, color = diet, shape = cohort)) +
  geom_point() +
  geom_label(aes(label = ferretID)) +
  geom_line(aes(group = ferretID)) +
  facet_grid(~diet+inf_route)+ 
  PlotTheme1 +
  DietcolScale
print(ContactSNVs_perFerret)
ggsave("ContactSNVs_perFerret.png", ContactSNVs_perFerret, path = savedir)
```

Do the number of variants correlate with titer or Ct?
```{r}
meta_small = select(meta,"ferretID","DPI","diet","inf_route","cohort","Ct_Mgene")
varcount = merge(gen_count_avg, meta_small, by = c("ferretID","DPI","diet","inf_route","cohort"))

ggplot(varcount, aes(x = snv_count, y = Ct_Mgene, color = diet)) +
  geom_point() +
  geom_smooth(method = "glm") +
  #ylim(20,30) +
  facet_grid(~inf_route) +
  PlotTheme1 +
  DietcolScale
# will get a warning when NAs are removed

ggplot(varcount, aes(x = snv_count, y = Ct_Mgene)) +
  geom_point() +
  geom_smooth(method = "glm") +
  #ylim(20,30) +
  facet_grid(~inf_route) +
  PlotTheme1 +
  DietcolScale
```

```{r}
titers = read.csv("FerretTiters.csv", header = TRUE, sep = ",")
titers_H9 = filter(titers, Virus == "A/Hong Kong/1073/1999 (H9N2)") %>%
  select(!c(Virus)) %>%
  droplevels()

varcount2 = merge(gen_count_avg,titers_H9, by = c("ferretID","DPI"))
varcount2 = varcount2[!duplicated(varcount2),]

var_titers = merge(varcount2, meta_small, by = c("ferretID","DPI","diet","inf_route","cohort"))
var_titers$LogTiter = log10(var_titers$Titer)
var_titers = var_titers[!duplicated(var_titers),]

var_titers$LogTiter = as.numeric(var_titers$LogTiter)
var_titers$snv_count = as.numeric(var_titers$snv_count)
```

```{r}
SNVs_Titer = ggplot(var_titers, aes(x = LogTiter, y = snv_count, color = diet)) +
  geom_point(aes(shape = DPI)) +
  geom_smooth(method = "glm") +
  PlotTheme1 +
  DietcolScale +
  xlim(0,7) +
  facet_grid(~inf_route)
print(SNVs_Titer)
ggsave("SNVs_Titer.pdf",SNVs_Titer,path = savedir, width = 8, height = 5)
ggsave("SNVs_Titer.png",SNVs_Titer,path = savedir, width = 8, height = 5)

# There are 16 points in the index ferrets with LogTiter < 3 in Index ferrets -> 2 samples (w/ 8 segments)
# Are these outliers that a skewing the data?

PossOutliers = filter(var_titers, inf_route == "Index" & diet == "Lean" & LogTiter < 3)
outlierSamps = levels(factor(PossOutliers$Sample))

var_titers_NoOut = var_titers %>% filter(!(Sample %in% outlierSamps))

SNVs_Titer_NoOut = ggplot(var_titers_NoOut, aes(x = LogTiter, y = snv_count, color = diet)) +
  geom_point(aes(shape = DPI)) +
  geom_smooth(method = "glm") +
  PlotTheme1 +
  DietcolScale +
  xlim(0,7) +
  facet_grid(~inf_route)
print(SNVs_Titer_NoOut)

SNVs_Titer_NoDiet = ggplot(var_titers, aes(x = LogTiter, y = snv_count)) +
  geom_point(aes(shape = DPI)) +
  geom_smooth(method = "glm") +
  PlotTheme1 +
  DietcolScale +
  xlim(0,7) +
  facet_grid(~inf_route)
print(SNVs_Titer_NoDiet)
```

```{r}
meta_smaller = select(meta_small,c(ferretID, DPI, diet, inf_route, cohort))
meta_smaller = meta_smaller[!duplicated(meta_smaller), ]

titers_H9_meta = merge(titers_H9, meta_smaller, 
                       by.x = c("ferretID", "DPI", "Cohort"), 
                       by.y = c("ferretID", "DPI", "cohort"))
titers_H9_meta = titers_H9_meta[!duplicated(titers_H9_meta), ]

titers_H9_meta = titers_H9_meta[!duplicated(titers_H9_meta), ]
titers_H9_meta$Titer[is.na(titers_H9_meta$Titer)] = 0
titers_H9_meta = filter(titers_H9_meta, Titer > 0)

AvgTiters = group_by(titers_H9_meta, diet, inf_route,DPI) %>%
  mutate(LogTiter = log10(Titer)) %>% 
  mutate(avgTiter = mean(LogTiter), sdTiter = sd(LogTiter))
AvgTiters$inf_route = factor(AvgTiters$inf_route, levels = c("Index","Contact"))

TiterPlot = ggplot(AvgTiters, aes(x = DPI, y = avgTiter, color = diet)) +
  geom_point() +
  geom_line(aes(group = diet)) +
  geom_errorbar(aes(ymin = avgTiter - sdTiter,
                    ymax = avgTiter + sdTiter)) +
  ylim(0,7) +
  facet_grid(~inf_route) +
  DietcolScale +
  PlotTheme1
print(TiterPlot)
ggsave("TiterPlot.png", TiterPlot, width = 7, height = 5, path = savedir)
```

Do the number of variants correlate with metabolic measures?
```{r}

```

SNV location plots
```{r}
SNVLocation = ggplot(minorvdf, aes(x = ntpos, y = ferretID)) +
  geom_point(aes(color = diet, shape = cohort)) +
  facet_grid(inf_route~segment) +
  PlotTheme1 +
  DietcolScale
print(SNVLocation)
ggsave(SNVLocation, file = "SNVLocation.pdf", path = savedir)
# ferret 1787 doesn't have any variants??

minorvdf$var = paste0(minorvdf$segment,"_",minorvdf$major,minorvdf$ntpos,minorvdf$minor)

# Comparing to SNVs found in the stock

stock = filter(minorvdf, DPI == "Stock")
stock = stock[!duplicated(stock), ]
stocksnv = levels(factor(stock$var))
length(stocksnv)

ferrets = filter(minorvdf, DPI != "Stock")
ferrets = ferrets[!duplicated(ferrets), ]

shared_w_stock = ferrets %>% filter(var %in% stocksnv)
nrow(shared_w_stock)
ferunique = ferrets %>% filter(!(var %in% stocksnv))
nrow(ferunique)
```
```{r}
stock_shared = stock[!duplicated(stock$var),] %>% ungroup() 
stock_shared = separate(stock_shared,segment, into = c("strain","CHROM"))
stock_shared$ntvar = paste0(stock_shared$major,stock_shared$ntpos,stock_shared$minor)
stock_shared$aavar = paste0(stock_shared$majoraa,stock_shared$aapos,stock_shared$minoraa)

stock_shared_smol = select(stock_shared,CHROM,ntvar,aavar) %>% droplevels()
```

SNV Location compared to stock
```{r}
StockSharedPlot = ggplot(shared_w_stock, aes(x = ntpos, y = ferretID)) +
  geom_point(aes(color = diet, shape = cohort), size = 2) +
  facet_grid(inf_route~segment, drop = FALSE) +
  PlotTheme1 +
  DietcolScale +
  ggtitle("SNVs found in stock")
print(StockSharedPlot)
ggsave(StockSharedPlot, file = "StockSharedPlot.pdf", height = 30, width = 15, path = savedir)

FerUniquePlot = ggplot(ferunique, aes(x = ntpos, y = ferretID)) +
  geom_point(aes(color = diet, shape = cohort)) +
  facet_grid(inf_route~segment) +
  PlotTheme1 +
  DietcolScale +
  ggtitle("SNVs not found in stock")
print(FerUniquePlot)
#ggsave(FerUniquePlot, file = "FerUniquePlot.pdf", path = savedir)
```
De novo SNVs
```{r}
denovo = ungroup(ferrets) %>% count(var)

filter(ferrets, var == "H9N2_PB2_A2214C") %>% ungroup() %>% count(ferretID,DPI)
filter(ferrets, var == "H9N2_PB2_A2214C") %>% count(minorfreq)
# found in basically every sample with a freq of 2-5% what is this
```
Venn diagram of obese and lean de novo SNVs
```{r}
o_var = filter(ferrets, diet == "Obese") 
o_var = unique(o_var$var)

l_var = filter(ferrets, diet == "Lean") 
l_var = unique(l_var$var)

diet_var <- list(Obese = o_var, Lean = l_var)

DietUniqueSNVS = ggVennDiagram(diet_var)
print(DietUniqueSNVS)
ggsave(DietUniqueSNVS, file = "DietUniqueSNVS.pdf", path = savedir)
```

Obese- and lean-specific SNVs
```{r}
lean = ferrets %>% 
  filter(var %in% l_var) %>% 
  filter(!(var %in% o_var)) 

lean$sample_var = paste0(lean$sample,"_",lean$var)
lean = lean[!duplicated(lean$sample_var),] 

lean$ferretID_var = paste0(lean$ferretID,"_",lean$var)
lean = lean[!duplicated(lean$ferretID_var),] 

lean = lean %>% 
  group_by(var) %>% 
  mutate(count = 1, totalsamp = sum(count))

obese = ferrets %>% 
  filter(var %in% o_var) %>% 
  filter(!(var %in% l_var)) 

obese$sample_var = paste0(obese$sample,"_",obese$var)
obese = obese[!duplicated(obese$sample_var),] 

obese$ferretID_var = paste0(obese$ferretID,"_",obese$var)
obese = obese[!duplicated(obese$ferretID_var),] 

obese = obese %>% 
  group_by(var) %>% 
  mutate(count = 1, totalsamp = sum(count))

dietunique = rbind(lean,obese)
dietunique = dietunique[!duplicated(dietunique), ] %>% droplevels()
dietunique = filter(dietunique, inf_route != "Aerosol")
  
DietUnique = ggplot(dietunique, aes(x = ntpos, y = segment)) +
  geom_point(aes(color = nonsyn, size = totalsamp)) + 
  ggtitle("Number of samples containing each variant - diet specific") +
  theme(legend.key = element_blank(),
         strip.background = element_rect(colour="black", fill="white"),
         axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  facet_grid(~diet+inf_route) +
  PlotTheme1
print(DietUnique)
ggsave(DietUnique, filename = "SegmentSNVPlot_DietUnqique.pdf", path = savedir)
```
```{r}
# make new version of this figure, separating out transmission v independent ferrets

```

```{r}
ggplot(filter(dietunique, totalsamp > 1), aes(x = DPI, y = minorfreq, color = nonsyn)) +
  geom_point() +
  geom_line(aes(group = ferretID)) +
  facet_grid(~diet+inf_route) +
  PlotTheme1 
```

```{r}
nonsyns = filter(dietunique, nonsyn == "nonsyn" & totalsamp > 1) %>% ungroup() %>% droplevels()
nonsyns = nonsyns[!duplicated(nonsyns$var),] 

nonsyns = separate(nonsyns,segment, into = c("strain","CHROM"))
nonsyns$ntvar = paste0(nonsyns$major,nonsyns$ntpos,nonsyns$minor)
nonsyns$aavar = paste0(nonsyns$majoraa,nonsyns$aapos,nonsyns$minoraa)

nonsyns_smol = select(nonsyns,CHROM,ntvar,aavar,diet,totalsamp) %>% droplevels()
write.csv(nonsyns_smol, "nonsyns.csv")
```

SNVs shared between diet groups
```{r}
shared = ferrets %>% 
  filter(var %in% o_var) %>% 
  filter(var %in% l_var) %>% 
  group_by(var) %>% 
  mutate(count = 1, totalsamp = sum(count))

SharedPlot = ggplot(shared, aes(x = ntpos, y = segment)) +
  geom_point(aes(size = totalsamp, color = nonsyn)) +
  ggtitle("Number of samples containing each variant - Shared between diet groups") +
  theme(legend.key = element_blank(),
         strip.background = element_rect(colour="black", fill="white"),
         axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  PlotTheme1
print(SharedPlot)
ggsave(SharedPlot, filename = "SegmentSNVPlot_DietShared.pdf", path = savedir)
```

```{r}
Minorfreq_dist = ggplot(ferrets, aes(x = minorfreq, fill = diet)) +
  geom_histogram(binwidth = 0.01) +
  PlotTheme1 +
  facet_grid(inf_route~diet) +
  DietcolScale_fill
print(Minorfreq_dist)
ggsave("Minorfreq_dist.pdf", Minorfreq_dist, path = savedir)
# obese seem to have fewer low-frequency de novo SNVs
```

```{r}
obese_index = filter(ferrets, diet == "Obese" & inf_route == "Index") %>% ungroup()
lean_index = filter(ferrets, diet == "Lean" & inf_route == "Index") %>% ungroup()
t.test(obese_index$minorfreq, lean_index$minorfreq)
# means are not different

obese_contact = filter(ferrets, diet == "Obese" & inf_route == "Contact") %>% ungroup()
lean_contact = filter(ferrets, diet == "Lean" & inf_route == "Contact") %>% ungroup()
t.test(obese_contact$minorfreq, lean_contact$minorfreq)
# means are not different

# QQ_Plot: compares the quantiles of two distributions, x =y suggests they are drawn from the same distribution
qqnorm(obese_index$minorfreq, main = "Obese Index - Test of Normal Distribution")
qqnorm(lean_index$minorfreq, main = "Lean Index - Test of Normal Distribution")
# neither distribution is normal
qqplot(obese_index$minorfreq,lean_index$minorfreq, xlab = "Obese Index", ylab = "Lean Index")

qqnorm(obese_contact$minorfreq, main = "Obese Contact - Test of Normal Distribution")
qqnorm(lean_contact$minorfreq, main = "Lean Contact - Test of Normal Distribution")
# neither distribution is normal
qqplot(obese_contact$minorfreq,lean_contact$minorfreq, xlab = "Obese Contact", ylab = "Lean Contact")

# Mann-Whitney-Wilcox test (Mann-Whitney U test): samples are not normally distributed and independent of each other
wilcox.test(obese_index$minorfreq,lean_index$minorfreq)
wilcox.test(obese_contact$minorfreq,lean_contact$minorfreq)
# distributions are not different

# Kolmogorov-Smirnov test: samples are not normally distributed and independent of each other
# "sensitive to differences in location and shape of the empirical CDFs of the two samples"
ks.test(obese_index$minorfreq,lean_index$minorfreq)
ks.test(obese_contact$minorfreq,lean_contact$minorfreq)
# distributions are not different
```

```{r}
highfreq = filter(ferrets, minorfreq > 0.25)
```

Muller Plots - looking at all minor variants over time

```{r}
#Muller plot with proportional minorfreq values
ferrets = group_by(ferrets, ferretID) %>% 
  mutate(proportion = minorfreq / sum(minorfreq))

xsamp = length(levels(factor(ferrets$ferretID)))
```

```{r}
#Muller plot with raw minorfreq values
Muller_absolute = ggplot(ferrets, aes(x=DPI, y=minorfreq, group = var, fill = var)) + 
  geom_area() +
  facet_grid(~ferretID) +
  PlotTheme1 +
  theme(legend.position = "none")
ggsave("Muller_absolute.pdf", Muller_absolute, width = xsamp*2, height = 8, limitsize = FALSE, path = savedir)

Muller_relative = ggplot(ferrets, aes(x=DPI, y=proportion, group = var, fill = var)) + 
  geom_area() +
  facet_grid(inf_route~ferretID) +
  PlotTheme1 +
  theme(legend.position = "none")
ggsave("Muller_relative.pdf", Muller_relative, width = xsamp*2, height = 8, limitsize = FALSE, path = savedir)
```
